flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java

```diff
@@ -282,97 +303,232 @@ public Status process() throws EventDeliveryException {
+   * transaction to ensure that data is replayed.
+   *
+   * @throws EventDeliveryException
    */
-  private GenericRecord deserialize(Event event, GenericRecord reuse)
-      throws EventDeliveryException {
-    decoder = DecoderFactory.get().binaryDecoder(event.getBody(), decoder);
-    // no checked exception is thrown in the CacheLoader
-    DatumReader<GenericRecord> reader = readers.getUnchecked(schema(event));
-    try {
-      return reader.read(reuse, decoder);
-    } catch (IOException ex) {
-      throw new EventDeliveryException("Cannot deserialize event", ex);
+  @VisibleForTesting
+  void closeWriter() throws EventDeliveryException {
+    if (writer != null) {
+      try {
+        writer.close();
+
+        long elapsedTimeSeconds = TimeUnit.MILLISECONDS.toSeconds(
+            System.currentTimeMillis() - lastRolledMillis);
+        LOG.info("Closed writer for {} after {} seconds and {} bytes parsed",
+            new Object[]{datasetUri, elapsedTimeSeconds, bytesParsed});
+      } catch (DatasetIOException ex) {
+        throw new EventDeliveryException("Check HDFS permissions/health. IO"
+            + " error trying to close the  writer for dataset " + datasetUri,
+            ex);
+      } catch (DatasetWriterException ex) {
+        throw new EventDeliveryException("Failure moving temp file.", ex);
+      } catch (RuntimeException ex) {
+        throw new EventDeliveryException("Error trying to close the  writer for"
+            + " dataset " + datasetUri, ex);
+      } finally {
+        // If we failed to close the writer then we give up on it as we'll
+        // end up throwing an EventDeliveryException which will result in
+        // a transaction rollback and a replay of any events written by this
+        // writer.
+        this.writer = null;
+        failurePolicy.close();
+      }
     }
   }
 
-  private static Schema schema(Event event) throws EventDeliveryException {
-    Map<String, String> headers = event.getHeaders();
-    String schemaURL = headers.get(
-        DatasetSinkConstants.AVRO_SCHEMA_URL_HEADER);
-    try {
-      if (headers.get(DatasetSinkConstants.AVRO_SCHEMA_URL_HEADER) != null) {
-        return schemasFromURL.get(schemaURL);
-      } else {
-        return schemasFromLiteral.get(
-            headers.get(DatasetSinkConstants.AVRO_SCHEMA_LITERAL_HEADER));
-      }
-    } catch (ExecutionException ex) {
-      throw new EventDeliveryException("Cannot get schema", ex.getCause());
+  /**
+   * Enter the transaction boundary. This will either begin a new transaction
+   * if one didn't already exist. If we're already in a transaction boundary,
+   * then this method does nothing.
+   *
+   * @param channel The Sink's channel
+   * @throws EventDeliveryException There was an error starting a new batch
+   *                                with the failure policy.
+   */
+  private void enterTransaction(Channel channel) throws EventDeliveryException {
```

Comment:
There's another way that the transaction can end up null when we would otherwise call commit. If you have commitOnBatch enabled then you close the transaction and set it to null during the commit. That means that shouldRoll can't check that the transaction is non-null but that means it will fail when it tries to commit the non-null transaction.

I could have two methods, one that checks the transaction and commits if it's open and another that has the precondition. I'd use the conditional version in the commit that happens during a writer roll and `stop()` and use the other method when committing a batch.

Or I can go back to having one method which check if the transaction is open.